File: feb-22-check-if-a-string-contains-all-binary-codes-of-size-k/solution.py
class Solution:
    def hasAllCodes(self, s: str, k: int) -> bool:
        n = len(s)
        need = 1 << k
        windows = n - k + 1
        if windows < need:
            return False

        seen = [False] * need
        mask = need - 1
        value = 0
        found = 0

        for i, ch in enumerate(s):
            value = ((value << 1) & mask) | (1 if ch == "1" else 0)
            if i >= k - 1 and not seen[value]:
                seen[value] = True
                found += 1
                if found == need:
                    return True

        return False

File: feb-22-check-if-a-string-contains-all-binary-codes-of-size-k/solution.java
class Solution {
    public boolean hasAllCodes(String s, int k) {
        int n = s.length();
        int need = 1 << k;
        int windows = n - k + 1;
        if (windows < need) {
            return false;
        }

        boolean[] seen = new boolean[need];
        int mask = need - 1;
        int value = 0;
        int found = 0;

        for (int i = 0; i < n; i++) {
            value = ((value << 1) & mask) | (s.charAt(i) - '0');
            if (i >= k - 1 && !seen[value]) {
                seen[value] = true;
                found++;
                if (found == need) {
                    return true;
                }
            }
        }

        return false;
    }
}
